### PR TITLE
Refactor logic related to tokens and fix few bugs

### LINKED ISSUE #2333: Set Backend JWT token for internal Key
### How will you implement it

---

### LINKED ISSUE #2364: Fix addtional todos in the source code for tenant mode JWT generation
$subject

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### LINKED ISSUE #2384: Error Appling API Key auth for petstore API
### Description:
Error Appling API Key auth for petstore API

```log
adapter_1   | 2021-11-03 17:45:03 INFO [notification_listener.go:79] - [messaging.handleNotification] [-] Event REMOVE_API_FROM_GATEWAY is received
adapter_1   | 2021-11-03 17:45:03 INFO [notification_listener.go:79] - [messaging.handleNotification] [-] Event DEPLOY_API_IN_GATEWAY is received
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:163] - [synchronizer.FetchAPIsFromControlPlane] [-] API 2653577f-2fb4-4066-887c-0f887f85af76 is added/updated to APIList for label [Default]
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:64] - [synchronizer.FetchAPIs] [-] Fetching APIs from Control Plane.
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:171] - [synchronizer.FetchAPIsFromControlPlane] [-] Pushing data to router and enforcer
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:58] - [synchronizer.PushAPIProjects] [-] Start Deploying 1 API/s...
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:91] - [synchronizer.PushAPIProjects] [-] Start deploying api from file (API_ID:REVISION_ID).zip : 2653577f-2fb4-4066-887c-0f887f85af76-f9073d74-6da6-4d5a-a3cf-4017c930d430.zip
adapter_1   | 2021-11-03 17:45:03 INFO [apis_impl.go:230] - [api.ApplyAPIProjectFromAPIM] [-] Deploying api SwaggerPetstore:1.0.5 in Organization carbon.super
adapter_1   | 2021-11-03 17:45:03 INFO [apis_impl.go:240] - [api.ApplyAPIProjectFromAPIM] [-] API SwaggerPetstore:1.0.5 with UUID "2653577f-2fb4-4066-887c-0f887f85af76" already deployed to vhost: localhost
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[default:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[petstore_auth:[write:pets read:pets]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[default:[write:pets read:pets]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 2. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:213] - [model.checkAPIKeyInOperationArray] [-] Inside security scheme &[{api_key api_key api_key } {default oauth2  } {api_key apiKey api_key header} {petstore_auth oauth2  }].
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 0. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 1. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 INFO [open_api.go:217] - [model.checkAPIKeyInOperationArray] [-] New method key 2. Value: map[api_key:[]]
adapter_1   | 2021-11-03 17:45:03 ERRO [apis_impl.go:222] - [api.ApplyAPIProjectFromAPIM.func1] [-] Recovered from panic. Error encountered while applying API SwaggerPetstore:1.0.5 to localhost.
adapter_1   | 2021-11-03 17:45:03 ERRO [apis_fetcher.go:105] - [synchronizer.PushAPIProjects] [-] Error occurred while applying project SwaggerPetstore:1.0.5 with UUID "2653577f-2fb4-4066-887c-0f887f85af76"
adapter_1   | 2021-11-03 17:45:03 INFO [apis_fetcher.go:113] - [synchronizer.PushAPIProjects] [-] Successfully deployed 1 API/s
```

### Steps to reproduce:
- Create API from https://petstore.swagger.io/v2/swagger.json
- Add API Key auth from runtime API Configs in APIM
- Deploy API
- check adapter logs

### Affected Product Version:
1.0.0-beta-snapshot

### Environment details (with versions):
- OS: mac
- Client: curl
- Env (Docker/K8s): docker compose

---

### LINKED ISSUE #2437: API key fails when enforcer debug enabled
### Description:
API-Key returns 401 when enforcer debug enabled via log4j2.properties

### Environment details (with versions):
- OS:
- Client:
- Env (Docker/K8s):

---

### PR DESCRIPTION
**Pull request by suksw**
- Move backend jwt generation logic to a single location
- Remove Reference holder and shorten subscription valdation logic
- Move common jwt and time related logic together
- Refactor apikey and internal key logic to reduce loop iterations, reuse methods.
- Add backend jwt generator to internal key flow with limited data
- Remove redundant null check for apikey because it isalready checked in canAuthenticate method
- update error response
- remove additional isVerified check to enable invalid key caching
- Refactor api-key apim flow and update test
- Remove app security optional check at api level
- Update header/query name to apikey in apim flow

### Issues
Fixes #2333 
Fixes #2384
Fixes #2364
Fixes #2437

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: samples/interceptors/certs/mg.pem [hunk 1]
```diff
@@ -0,0 +1,21 @@
+-----BEGIN CERTIFICATE-----
```
Comment (renuka-fernando) [inline samples/interceptors/certs/mg.pem:1]: Is this file insertion intentional?

Comment (renuka-fernando) [reply samples/interceptors/certs/mg.pem:1]: The whole dir `samples/interceptors`

Comment (suksw) [reply samples/interceptors/certs/mg.pem:1]: No. Will be rebasing after removing all of them.

File: integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml [hunk 2]
```diff
@@ -114,86 +62,17 @@ definitions:
...
-        description: User Status
-    xml:
-      name: User
-externalDocs:
-  description: Find out more about Swagger
-  url: 'http://swagger.io'
-x-wso2-application-security:
-  security-types:
-    - api_key
+
```
Comment (pubudu538) [inline integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml:72]: Let's remove new lines.

File: integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml [hunk 3]
```diff
@@ -114,86 +62,17 @@ definitions:
...
+
+
+
+
+
+
```
Comment (pubudu538) [inline integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml:78]: Let's remove new lines.

File: adapter/internal/oasparser/model/swagger.go [hunk 4]
```diff
@@ -175,27 +175,6 @@ func setResourcesSwagger(swagger2 spec.Swagger, securitySchemes *[]SecuritySchem
 // Sets security definitions defined in swagger 2 format.
 func setSecurityDefinitions(swagger2 spec.Swagger) []SecurityScheme {
 	var securitySchemes []SecurityScheme
-	var isApplicationSecurityOptional = true
-	result, ok := swagger2.Extensions[xWso2ApplicationSecurity].(map[string]interface{})
-	if ok {
-		for key, value := range result {
-			if key == Optional && value != true {
-				isApplicationSecurityOptional = false
```
Comment (suksw) [inline adapter/internal/oasparser/model/swagger.go:183]: https://github.com/wso2/product-microgateway/blob/e20020a77289e3bb501085d05df1422b985fca95/enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/AuthFilter.java#L62

File: adapter/internal/discovery/xds/server.go [hunk 5]
```diff
@@ -270,14 +270,14 @@ func UpdateAPI(vHost string, apiProject mgw.ProjectAPI, environments []string) (
...
-				schemes = append(schemes, model.SecurityScheme{Type: value})
+		schemes = mgwSwagger.GetSecurityScheme()
+		for _, value := range apiYaml.SecurityScheme {
+			if value == model.APIKeyInAppLevelSecurity {
+				schemes = append(schemes, model.SecurityScheme{DefinitionName: model.APIKeyInAppLevelSecurity,
+					Type: value, Name: model.APIKeyNameWithApim})
 			}
-			mgwSwagger.SetSecurityScheme(schemes)
 		}
-		mgwSwagger.SetXWso2AuthHeader(apiYaml.AuthorizationHeader)
```
Comment (suksw) [inline adapter/internal/discovery/xds/server.go:280]: Add this back

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/InternalAPIKeyAuthenticator.java [hunk 6]
```diff
@@ -147,37 +157,18 @@ public AuthenticationContext authenticate(RequestContext requestContext) throws
...
-                        throw new APISecurityException(APIConstants.StatusCodes.UNAUTHENTICATED.getCode(),
-                                APISecurityConstants.API_AUTH_INVALID_CREDENTIALS,
-                                APISecurityConstants.API_AUTH_INVALID_CREDENTIALS_MESSAGE);
-                    } finally {
-                        if (Utils.tracingEnabled()) {
-                            verifyTokenWithoutCacheSpanScope.close();
-                            Utils.finishSpan(verifyTokenWithoutCacheSpan);
-                        }
+                    isVerified = verifyTokenWhenNotInCache(jwsHeader, signedJWT, splitToken, payload, "InternalKey");
+                    if (Utils.tracingEnabled()) {
```
Comment (suksw) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/InternalAPIKeyAuthenticator.java:169]: Add inside a finally block

Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/InternalAPIKeyAuthenticator.java]: fixed with https://github.com/wso2/product-microgateway/pull/2432/commits/c96313affe6a7144b4c8daa0d2f7c1d35d61a2a9

File: adapter/internal/discovery/xds/server.go (hunk 5 above)
Comment (suksw) [reply adapter/internal/discovery/xds/server.go:280]: fixed with https://github.com/wso2/product-microgateway/pull/2432/commits/c96313affe6a7144b4c8daa0d2f7c1d35d61a2a9

File: integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml (hunk 3 above)
Comment (suksw) [reply integration/test-integration/src/test/resources/openAPIs/app_level_security_openAPI.yaml]: fixed with https://github.com/wso2/product-microgateway/pull/2432/commits/c96313affe6a7144b4c8daa0d2f7c1d35d61a2a9

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java [hunk 7]
```diff
@@ -83,116 +86,129 @@ public boolean canAuthenticate(RequestContext requestContext) {
...
-        if ((isAppLevelSecurityRequest && StringUtils.isEmpty(apiKey)) || (StringUtils.isEmpty(apiKey) &&
-                apiKeyLocation.equals(APIConstants.SWAGGER_API_KEY_IN_QUERY))) {
-            Map<String, String> queryParameters = requestContext.getQueryParameters();
-            apiKey = getAPIKeyFromMap(queryParameters, apiKeyName);
+        if (isAppLevelApiKeySecurityEnabled) {
+            // If API key enabled at Application level via APIM, key can exist in header or query param
+            String apiKeyName = requestContext.getMatchedAPI().
+                    getSecuritySchemeDefinitions().get(APIConstants.API_SECURITY_API_KEY).getName();
+            if (requestContext.getHeaders().containsKey(apiKeyName)) {
+                log.info("getHeaders");
```
[the remark this answers is not in the document]
Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/jwt/APIKeyAuthenticator.java]: fixed with https://github.com/wso2/product-microgateway/pull/2432/commits/c96313affe6a7144b4c8daa0d2f7c1d35d61a2a9

**Review by pubudu538 — APPROVED** [review]

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/KeyValidator.java [hunk 8]
```diff
@@ -127,31 +106,18 @@ public boolean validateScopes(TokenValidationContext validationContext) throws E
...
 
-    private boolean validateSubscriptionDetails(String uuid, String context, String version, String consumerKey,
-            String keyManager, APIKeyValidationInfoDTO infoDTO) throws EnforcerException {
-        boolean defaultVersionInvoked = false;
-        String apiTenantDomain = FilterUtils.getTenantDomainFromRequestURL(context);
-        if (apiTenantDomain == null) {
-            apiTenantDomain = APIConstants.SUPER_TENANT_DOMAIN_NAME;
+    public static APIKeyValidationInfoDTO validateSubscription(String uuid, String apiContext, String apiVersion,
+                                                         String consumerKey, String keyManager) {
+        if (log.isDebugEnabled()) {
```
Comment (praminda) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/KeyValidator.java:111]: can you remove the isDebugEnabled check and log in templated logging mode?

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+/**
+ * Contains Util methods related to backend JWT generation
```
Comment (praminda) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java:47]: Add a fullstop after the first sentence of javadoc as a practice. there are checkstyle rules catching this as an style error
```suggestion
 * Contains Util methods related to backend JWT generation.
```

Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java:47]: I build the pack locally before committing. Must have been in the warning state. will check other places as well.

Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java]: Fixed with https://github.com/wso2/product-microgateway/pull/2432/commits/21dd29c99c97e3e87c32e09f0a24f4b51be5fbf8

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/KeyValidator.java (hunk 8 above)
Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/security/KeyValidator.java]: When testing after making the changes I came across this https://github.com/wso2/product-microgateway/issues/2437. Tried adding templates to all the debug logs but it did not stop. Therefore removed some of the debug logs via this https://github.com/wso2/product-microgateway/pull/2432/commits/21dd29c99c97e3e87c32e09f0a24f4b51be5fbf8 which fixed the issue. The remaining changes are here. https://github.com/wso2/product-microgateway/pull/2432/commits/4ff61c3dac7e00bd0d2d4ad2ad32e6fee5daca6c

**Review by praminda — APPROVED** [review]

File: enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    /**
+     * Get Time to live
+     * @return -1L * 1000
+     */
+    public static long getTTL() {
+        return ttl * 1000;
+    }
+
+    /**
+     * Generates or gets the Cached Backend JWT token
```
Comment (praminda) [inline enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java:62]: you have to make this change in all places
```suggestion
     * Generates or gets the Cached Backend JWT token.
```

Comment (suksw) [reply enforcer-parent/enforcer/src/main/java/org/wso2/choreo/connect/enforcer/util/BackendJwtUtils.java:62]: Fixed with https://github.com/wso2/product-microgateway/pull/2442

**Pull request merged**